### PR TITLE
[LoongArch] Modify expanding code sequence for PseudoLA_TLS_LE

### PR DESCRIPTION
**Pull request by zhaoqi5**
Before this commit, PseudoLA_TLS_LE for normal/medium code model expand normally to:
```
  lu12i.w $rd, %le_hi20(sym)
  ori $rd, $rd, %le_lo12(sym)
```

This commit changes the result to:
```
  lu12i.w $rd, %le_hi20_r(sym)
  add.w/d $rd, $rd, $tp, %le_add_r(sym)
  addi.w/d $rd, $rd, %le_lo12_r(sym)
```

This aims to be optimized by linker relaxation in the future.

This commit makes no change to PseudoLA_TLS_LE in large code model.

### CODE REVIEW

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp [hunk 1]
```diff
@@ -1819,6 +1818,10 @@ SDValue LoongArchTargetLowering::getStaticTLSAddr(GlobalAddressSDNode *N,
   SDValue Offset = Large
                        ? SDValue(DAG.getMachineNode(Opc, DL, Ty, Tmp, Addr), 0)
                        : SDValue(DAG.getMachineNode(Opc, DL, Ty, Addr), 0);
+
+  if (IsNormalOrMediumLE)
```
Comment (wangleiat) [inline llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp:1822]:
```suggestion
  if (Opc == LoongArch::PseudoLA_TLS_LE && !Large)
```

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.h [hunk 2]
```diff
@@ -301,7 +301,8 @@ class LoongArchTargetLowering : public TargetLowering {
   SDValue getAddr(NodeTy *N, SelectionDAG &DAG, CodeModel::Model M,
                   bool IsLocal = true) const;
   SDValue getStaticTLSAddr(GlobalAddressSDNode *N, SelectionDAG &DAG,
-                           unsigned Opc, bool UseGOT, bool Large = false) const;
+                           unsigned Opc, bool UseGOT, bool Large = false,
+                           bool IsNormalOrMediumLE = false) const;
```
Comment (wangleiat) [inline llvm/lib/Target/LoongArch/LoongArchISelLowering.h:305]: This parameter should be removable.
`IsNormalOrMediumLE` = `(Opc == LoongArch::PseudoLA_TLS_LE && !Large)`

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp (hunk 1 above)
Comment (wangleiat) [reply llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp:1822]: It's better to add a comment here. If it is LE, the `add tp` operation will occur during the pseudo-instruction expansion.

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.h (hunk 2 above)
Comment (zhaoqi5) [reply llvm/lib/Target/LoongArch/LoongArchISelLowering.h:305]: When TLS_LE calling `getStaticTLSAddr()`, parameter `Large` always uses the default value `false` no matter what the actual codemodel is. So `LoongArch::PseudoLA_TLS_LE && !Large` is always `true`. But we only expect `IsNormalOrMediumLE` to be true when the codemodel is actually normal or medium.

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp (hunk 1 above)
Comment (zhaoqi5) [reply llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp:1822]: Thanks. I will add this.

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.h (hunk 2 above)
Comment (wangleiat) [reply llvm/lib/Target/LoongArch/LoongArchISelLowering.h:305]: Sorry, I didn’t notice that. In that case, can we actually pass Large here and handle different Opc values accordingly?

Comment (zhaoqi5) [reply llvm/lib/Target/LoongArch/LoongArchISelLowering.h]: I have modified according to your suggestions. Thanks.

**Review by wangleiat — APPROVED** [review]
LGTM, thanks.

**Pull request merged**